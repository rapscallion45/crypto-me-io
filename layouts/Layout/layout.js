import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import Particles from 'react-tsparticles';
import { loadFull } from 'tsparticles';
import CssBaseline from '@mui/material/CssBaseline';
import Box from '@mui/material/Box';
import AppBar from '@mui/material/AppBar';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import Meta from '../../components/meta';
import Footer from '../../components/footer';
import Logo from '../../components/logo';
import Link from '../../components/link';
import Loader from '../../components/loader';
import MHidden from '../../components/m-hidden';
import conf from '../../utils/particlesConf';
import userActions from '../../redux/actions/actions';

const Layout = function Layout({ children }) {
  const dispatch = useDispatch();
  const currencyData = useSelector((state) => state.globalCurrencyData);

  useEffect(() => {
    dispatch(userActions.getGlobalCurrencyData());
  }, []);

  const particlesInit = async (main) => {
    await loadFull(main);
  };

  const particlesLoaded = () => {};

  return (
    <>
      <Meta />
      <Box
        sx={{
          display: 'flex',
          flexDirection: 'column',
          minHeight: '100vh',
          background:
            'linear-gradient(90deg, rgba(2,0,36,1) 0%, rgba(27,25,180,1) 0%, rgba(31,31,212,1) 45%, rgba(0,212,255,1) 100%)',
        }}
      >
        <CssBaseline />
        <Box sx={{ flexGrow: 1, maxHeight: 80 }}>
          <AppBar position="sticky" color="secondary">
            <Toolbar variant="dense">
              <Link href="/" sx={{ mr: 5 }}>
                <Logo />
              </Link>
              <MHidden width="lgDown">
                <Loader
                  dataLoaded={currencyData.loaded}
                  dataLoading={currencyData.loading}
                  dataError={currencyData.error}
                >
                  <Typography variant="body2" color="text.primary">
                    Coins:{' '}
                    <Link
                      variant="button"
                      color="primary.main"
                      href="/"
                      sx={{ my: 1, fontWeight: '900' }}
                    >
                      {currencyData.data?.active_cryptocurrencies}
                    </Link>
                  </Typography>
                  <Typography ml={1} variant="body2" color="text.primary">
                    Markets:{' '}
                    <Link
                      variant="button"
                      color="primary.main"
                      href="/"
                      sx={{ my: 1, fontWeight: '900' }}
                    >
                      {currencyData.data?.markets}
                    </Link>
                  </Typography>
                  <Typography ml={1} variant="body2" color="text.primary">
                    Coins:{' '}
                    <Link
                      variant="button"
                      color="primary.main"
                      href="/"
                      sx={{ my: 1, fontWeight: '900' }}
                    >
                      {currencyData.data?.active_cryptocurrencies}
                    </Link>
                  </Typography>
                  <Typography ml={1} mr={6} variant="body2" color="text.primary">
                    Top Market Share:{' '}
                    <Link
                      variant="button"
                      color="primary.main"
                      href="/"
                      sx={{ my: 1, fontWeight: '900' }}
                    >
                      {currencyData.data?.active_cryptocurrencies}
                    </Link>
                  </Typography>
                </Loader>
                <nav>
                  <Link
                    variant="button"
                    color="primary.main"
                    href="/"
                    sx={{ my: 1, mx: 1.5, fontWeight: '900', textTransform: 'uppercase' }}
                  >
                    Markets
                  </Link>
                </nav>
              </MHidden>
              <MHidden width="lgUp">
                <IconButton aria-label="open-menu">
                  <MenuIcon color="primary" sx={{ fontSize: 30 }} />
                </IconButton>
              </MHidden>
            </Toolbar>
          </AppBar>
        </Box>
        {children}
        <Particles id="tsparticles" init={particlesInit} loaded={particlesLoaded} options={conf} />
        <Footer />
      </Box>
    </>
  );
};

Layout.propTypes = {
  children: PropTypes.node.isRequired,
};

export default Layout;
