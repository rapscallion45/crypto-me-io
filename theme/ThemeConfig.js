import PropTypes from 'prop-types';
import { useMemo } from 'react';
import { CacheProvider } from '@emotion/react';
import { CssBaseline } from '@mui/material';
import { ThemeProvider, createTheme } from '@mui/material/styles';
import shape from './shape';
import palette from './palette';
import typography from './typography';
import componentsOverride from './overrides';
import shadows, { customShadows } from './shadows';

const ThemeConfig = function Themeconfig({ emotionCache, children }) {
  const themeOptions = useMemo(
    () => ({
      palette,
      shape,
      typography,
      shadows,
      customShadows,
    }),
    []
  );

  const theme = createTheme(themeOptions);
  theme.components = componentsOverride(theme);

  return (
    <CacheProvider value={emotionCache}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </CacheProvider>
  );
};

ThemeConfig.propTypes = {
  children: PropTypes.node,
};

export default ThemeConfig;
