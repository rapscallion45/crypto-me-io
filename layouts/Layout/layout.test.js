import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import router from 'next/router';
import createEmotionCache from '../../utils/createEmotionCache';
import ThemeConfig from '../../theme/ThemeConfig';
import rootReducer from '../../redux/reducers/reducers';
import Layout from './layout';

const middleware = [thunkMiddleware];

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/user',
      pathname: '',
      query: { id: 7 },
      asPath: '',
      push: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    };
  },
}));

describe('Layout', () => {
  it('Renders correctly enzyme', () => {
    /* Arrange */
    const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));
    const TestChild = function TestChild() {
      return <div>Test Child</div>;
    };
    const useRouter = jest.spyOn(router, 'useRouter');
    useRouter.mockImplementation(() => ({
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
      push: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    }));

    /* Act */
    const wrapper = shallow(
      <Provider store={testStore}>
        <ThemeConfig emotionCache={createEmotionCache()}>
          <Layout>
            <TestChild />
          </Layout>
        </ThemeConfig>
      </Provider>
    );

    /* Assert */
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('Elements', () => {
    it('Should render link to home page', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));
      const TestChild = function TestChild() {
        return <div>Test Child</div>;
      };
      const useRouter = jest.spyOn(router, 'useRouter');
      useRouter.mockImplementation(() => ({
        route: '/',
        pathname: '',
        query: '',
        asPath: '',
        push: jest.fn(),
        events: {
          on: jest.fn(),
          off: jest.fn(),
        },
        beforePopState: jest.fn(() => null),
        prefetch: jest.fn(() => null),
      }));

      /* Act */
      const wrapper = render(
        <Provider store={testStore}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <Layout>
              <TestChild />
            </Layout>
          </ThemeConfig>
        </Provider>
      );

      /* Assert */
      expect(wrapper.getByRole('link', { name: 'logo-image' })).toBeInTheDocument();
    });

    it('Should render passed children components', () => {
      /* Arrange */
      const testStore = createStore(rootReducer, {}, applyMiddleware(...middleware));
      const TestChild = function TestChild() {
        return <div>Test Child</div>;
      };
      const useRouter = jest.spyOn(router, 'useRouter');
      useRouter.mockImplementation(() => ({
        route: '/',
        pathname: '',
        query: '',
        asPath: '',
        push: jest.fn(),
        events: {
          on: jest.fn(),
          off: jest.fn(),
        },
        beforePopState: jest.fn(() => null),
        prefetch: jest.fn(() => null),
      }));

      /* Act */
      const wrapper = render(
        <Provider store={testStore}>
          <ThemeConfig emotionCache={createEmotionCache()}>
            <Layout>
              <TestChild />
            </Layout>
          </ThemeConfig>
        </Provider>
      );

      /* Assert */
      expect(wrapper.getByText('Test Child')).toBeInTheDocument();
    });
  });
});
