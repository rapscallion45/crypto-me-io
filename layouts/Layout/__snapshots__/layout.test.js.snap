// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Renders correctly enzyme 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "onStateChange": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Themeconfig
    emotionCache={
      Object {
        "insert": [Function],
        "inserted": Object {},
        "key": "css",
        "nonce": undefined,
        "registered": Object {},
        "sheet": StyleSheet {
          "_insertTag": [Function],
          "before": null,
          "container": <head />,
          "ctr": 0,
          "insertionPoint": undefined,
          "isSpeedy": false,
          "key": "css",
          "nonce": undefined,
          "prepend": undefined,
          "tags": Array [],
        },
      }
    }
  >
    <Layout>
      <TestChild />
    </Layout>
  </Themeconfig>
</ContextProvider>
`;
